samza-log4j/src/main/java/org/apache/samza/logging/log4j/StreamAppender.java

```diff
@@ -206,13 +257,53 @@ protected void setupSystem() {
 
     setSerde(log4jSystemConfig, systemName, streamName);
 
-    systemProducer = systemFactory.getProducer(systemName, config, new MetricsRegistryMap());
+    systemProducer = systemFactory.getProducer(systemName, config, metricsRegistry);
     systemStream = new SystemStream(systemName, streamName);
     systemProducer.register(SOURCE);
     systemProducer.start();
 
     log.info(SOURCE + " has been registered in " + systemName + ". So all the logs will be sent to " + streamName
         + " in " + systemName + ". Logs are partitioned by " + key);
+
+    startQueueToProducerTransferThread();
+  }
+
+  private void startQueueToProducerTransferThread() {
+
+    try {
+      // Serialize the key once, since we will use it for every event.
+      final byte[] keyBytes = key.getBytes("UTF-8");
+
+      Runnable logQueueConsumer = () -> {
+        while (true) {
+          try {
+            byte[] serializedLogEvent = logQueue.take();
+
+            OutgoingMessageEnvelope outgoingMessageEnvelope =
+                new OutgoingMessageEnvelope(systemStream, keyBytes, serializedLogEvent);
+            systemProducer.send(SOURCE, outgoingMessageEnvelope);
+
+            if (Thread.interrupted()) {
+              break;
+            }
+          } catch (InterruptedException e) {
+            break; // Someone wants us to stop, so stop.
```

Comment:
Can just invoke a `Thread.currentThread().interrupt()` and let the loop's while condition be, `while(!Thread.currentThread().isInterrupted())` instead of a `while(true)`